File: e6/java/Solution2.java
class Solution2 {
    /**
     * @param A and B: sorted integer array A and B.
     * @return: A new sorted integer array
     */
    public int[] mergeSortedArray(int[] A, int[] B) {
        // Write your code here
		int[] shortArray, longArray;
		shortArray = A.length < B.length ? A : B;
		longArray = A.length >= B.length ? A : B;
		
		int newbegin = 0;
		int[] result = new int[A.length+B.length];
		int resultIdx = 0;
		for (int i=0; i<shortArray.length; i++) {
			int tmp = shortArray[i];
			if (tmp <= longArray[newbegin]) {
				result[resultIdx] = tmp;
				resultIdx++;
			}
			else if (tmp >= longArray[longArray.length-1]) {
				/*
				System.out.println(" tmp > longArray.max");
				System.out.println(newbegin);
				System.out.println(resultIdx);
				System.out.println(A.length+B.length-resultIdx);
				*/
				System.arraycopy(longArray, newbegin, result, resultIdx, longArray.length - newbegin);
				System.arraycopy(shortArray, i, result, resultIdx+longArray.length-newbegin, shortArray.length - i);
				newbegin = longArray.length;
				resultIdx = A.length+B.length;
				break;
			}
			else {
				System.out.println(" tmp between longArray.min and longArray.max");
				int idx = find(longArray, tmp);
				/*
				System.out.println(newbegin);
				System.out.println(resultIdx);
				System.out.println(idx-newbegin);
				*/
				System.arraycopy(longArray, newbegin, result, resultIdx, idx-newbegin);
				resultIdx += idx-newbegin;
				newbegin = idx;
				result[resultIdx++] = tmp;
			}
		}
        if (newbegin < longArray.length)
			System.arraycopy(longArray, newbegin, result, resultIdx, longArray.length-newbegin);
		return result;
    }

	public static int find(int[] array, int e) {
		if (e <= array[0])
			return 0;
		if (e >= array[array.length-1])
			return array.length;
		int lp = 0;
		int rp = array.length - 1;
		while (rp -lp > 1) {
			int mid = (lp + rp) / 2;
			if (array[mid] > e) {
				rp = mid;
			}
			else {
				lp = mid;
			}
		}
		return rp;
	}

	public static void main(String[] args) {
		int[] a = {1,3,5};
		int[] b = {1,2,3,4,5};
		int[] newArray = new Solution2().mergeSortedArray(a, b);
		for (int i:newArray) {
			System.out.println(i);
		}
	}
}

File: e6/python/solution2.py
import sys
class Solution:
    #@param A and B: sorted integer array A and B.
    #@return: A new sorted integer array
    def mergeSortedArray(self, A, B):
        if len(A) > len(B):
            shortArray = B
            longArray = A
        else:
            shortArray = A
            longArray = B
        newbegin = 0
        ret = []
        for i in range(len(shortArray)):
            e = shortArray[i]
            if e <= longArray[newbegin]:
                ret.append(e)
            elif e >= longArray[-1]:
                ret += longArray[newbegin:]
                ret += shortArray[i:]
                newbegin = len(longArray)
                break
            else:
                tmp = self.find(longArray, e)
                ret += longArray[newbegin:tmp]
                ret.append(e)
                newbegin = tmp

        if newbegin < len(longArray):
            ret += longArray[newbegin:]
        return ret


    def find(self, array, element):
        #make sure element  between min and max of array
        if element <= array[0]:
            return 0
        if element >= array[-1]:
            return len(array)
        lp = 0
        rp = len(array) - 1
        while rp - lp > 1:
            if array[(lp + rp) / 2] > element:
                rp = (lp + rp) / 2
            else:
                lp = (lp + rp) / 2
        return rp


def main(argv=None):
    if argv is None:
        argv = sys.argv
    lista = range(1, 10000000)
    #lista = range(0, 10)
    listb = [1,3,5]
    Solution().mergeSortedArray(lista, listb)


if __name__ == '__main__':
    main()
